File: 01-游戏开发/01-服务端/网络编程/Reactor模式/python/reactor.py
"""
reactor: single-threaded event-driven game server using selectors.

The reactor is a small framework that decouples:
  - Synchronous Event Demultiplexer (selectors) -> block until fd ready
  - Event Handler API: handle_accept / handle_read / handle_write
  - Reactor core: register/deregister + event loop

This demo builds a tiny "chat server" skeleton: each connection has
a session object holding the half-read buffer; handlers are pluggable.
    python3 reactor.py 9000
"""
import selectors
import socket
import sys
import threading

sel = selectors.DefaultSelector()


# ---- Reactable handler interface ----
class Handler:
    def handle_accept(self, listener): ...
    def handle_read(self, sess): ...
    def handle_close(self, sess): ...


# ---- Default echo handler ----
class EchoHandler(Handler):
    def handle_accept(self, listener):
        s, _ = listener.accept()
        s.setblocking(False)
        sess = Session(s, self)
        sel.register(s, selectors.EVENT_READ, data=sess)

    def handle_read(self, sess):
        data = sess.sock.recv(4096)
        if not data:
            self.handle_close(sess)
            return
        sess.feed(data)              # session implements the receive buffer
        sess.sendall(b"echo:" + data)  # naive full send (no EAGAIN split)

    def handle_close(self, sess):
        sel.unregister(sess.sock)
        sess.sock.close()


class Session:
    def __init__(self, sock: socket.socket, handler: Handler):
        self.sock = sock
        self.handler = handler
        self.buf = bytearray()

    def feed(self, data: bytes):
        # The session receives bytes that may represent one or more
        # application-layer messages; a real reactor would dispatch them
        # here (see demo 119 TCP framing).
        self.buf.extend(data)

    def sendall(self, data: bytes):
        # Non-blocking-ish send: try once; if EAGAIN, the reactor would
        # switch to EPOLLOUT mode. Skipped for brevity.
        self.sock.sendall(data)


# ---- Reactor core ----
class Reactor:
    def __init__(self, sel_impl=selectors.DefaultSelector()):
        self.sel = sel_impl
        self.handler: Handler = EchoHandler()
        self.keep_running = True

    def add_listener(self, port: int):
        l = socket.socket()
        l.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        l.bind(("0.0.0.0", port))
        l.listen(16)
        l.setblocking(False)
        # Tag the listener with a Handler binding so we know to accept
        self.sel.register(l, selectors.EVENT_READ, data=("listener", l))
        return l

    def run(self, port: int = 9000):
        self.add_listener(port)
        print(f"reactor listening on :{port} (sel={type(self.sel).__name__})")
        while self.keep_running:
            for key, mask in self.sel.select(timeout=1.0):
                if key.data[0] == "listener":
                    self.handler.handle_accept(key.data[1])
                else:
                    sess: Session = key.data
                    try:
                        self.handler.handle_read(sess)
                    except ConnectionResetError:
                        self.handler.handle_close(sess)


if __name__ == "__main__":
    p = int(sys.argv[1]) if len(sys.argv) > 1 else 9000
    Reactor().run(p)

File: 01-游戏开发/01-服务端/网络编程/Reactor模式/c/reactor.c
/*
 * reactor: single-threaded event-driven game server framework.
 * Built on epoll (Linux), with a clean reactor pattern API:
 *   - register/deregister a fd with the reactor
 *   - associate any user-defined state (e.g. player session)
 *   - dispatch read/write events to user handlers
 *
 * The demo registers a listening socket + echo handlers and reuses this
 * pattern to echo everything back to the client.
 *
 *   gcc -O2 -Wall -Wextra reactor.c -o reactor
 *   ./reactor 9000
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/epoll.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#define MAX_EVENTS 64
#define BUF_SIZE 4096

/* Forward declaration of session struct used by user handlers */
struct session {
    int fd;
    int listen;
    char buf[BUF_SIZE];
    int len;
    int closed;
};

/* ---- Reactor API ----
 * The user handler is called whenever:
 *   - fd has EPOLLIN  and read_handler must drain the socket
 *   - fd has EPOLLOUT and write_handler may try sending
 * Handlers must be non-blocking; if no work, return without blocking.
 */
typedef struct reactor {
    int epfd;
    /* Simple fd -> session lookup (real engine: hash map) */
    struct session *sessions[1024];
} reactor_t;

static int reactor_init(reactor_t *r) {
    r->epfd = epoll_create1(0);
    if (r->epfd < 0) return -1;
    memset(r->sessions, 0, sizeof(r->sessions));
    return 0;
}

static int reactor_register(reactor_t *r, int fd, uint32_t events,
                            struct session *s) {
    if (fd < 0 || fd >= 1024) return -1;
    r->sessions[fd] = s;
    struct epoll_event ev = { .events = events, .data.ptr = s };
    return epoll_ctl(r->epfd, EPOLL_CTL_ADD, fd, &ev);
}

static int reactor_modify(reactor_t *r, int fd, uint32_t events,
                          struct session *s) {
    struct epoll_event ev = { .events = events, .data.ptr = s };
    return epoll_ctl(r->epfd, EPOLL_CTL_MOD, fd, &ev);
}

static int reactor_unregister(reactor_t *r, int fd, struct session *s) {
    s->closed = 1;
    r->sessions[fd] = NULL;
    return epoll_ctl(r->epfd, EPOLL_CTL_DEL, fd, NULL);
}

/* ---- User-defined handlers ---- */

static void on_accept(reactor_t *r, struct session *listen_s) {
    for (;;) {
        int cfd = accept4(listen_s->fd, NULL, NULL, SOCK_NONBLOCK);
        if (cfd < 0) {
            if (errno == EAGAIN || errno == EWOULDBLOCK) return;
            perror("accept4"); return;
        }
        struct session *c = calloc(1, sizeof(*c));
        c->fd = cfd;
        c->listen = 0;
        reactor_register(r, cfd, EPOLLIN, c);
        printf("accept fd=%d\n", cfd);
    }
}

static void on_read(reactor_t *r, struct session *s) {
    ssize_t n = recv(s->fd, s->buf + s->len, sizeof(s->buf) - s->len, 0);
    if (n == 0) {
        printf("peer closed fd=%d\n", s->fd);
        close(s->fd);
        reactor_unregister(r, s->fd, s);
        free(s);
        return;
    }
    if (n < 0) {
        if (errno != EAGAIN) perror("recv");
        return;
    }
    s->len += (int)n;
    /* Best-effort echo: in production, parse the buffer against a protocol */
    ssize_t sent = send(s->fd, s->buf, s->len, 0);
    if (sent > 0) {
        if (sent < s->len) {
            memmove(s->buf, s->buf + sent, s->len - sent);
            s->len -= (int)sent;
        } else {
            s->len = 0;
        }
    }
}

int main(int argc, char *argv[]) {
    int port = (argc > 1) ? atoi(argv[1]) : 9000;

    /* listen socket */
    int lfd = socket(AF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
    int on = 1;
    setsockopt(lfd, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
    struct sockaddr_in addr = { 0 };
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_ANY);
    addr.sin_port = htons(port);
    if (bind(lfd, (struct sockaddr *)&addr, sizeof(addr)) < 0 ||
        listen(lfd, 16) < 0) { perror("listen"); return 1; }

    reactor_t reactor;
    reactor_init(&reactor);

    struct session *listen_s = calloc(1, sizeof(*listen_s));
    listen_s->fd = lfd;
    listen_s->listen = 1;
    reactor_register(&reactor, lfd, EPOLLIN, listen_s);

    printf("reactor listening on :%d (epfd=%d)\n", port, reactor.epfd);

    /* Reactor main loop:
     *   Phase 1: synchronous event demultiplexer
     *   Phase 2: dispatch to handlers
     */
    struct epoll_event events[MAX_EVENTS];
    for (;;) {
        int nev = epoll_wait(reactor.epfd, events, MAX_EVENTS, 1000);
        if (nev < 0) { if (errno == EINTR) continue; perror("epoll_wait"); break; }
        for (int i = 0; i < nev; i++) {
            struct session *s = (struct session *)events[i].data.ptr;
            uint32_t e = events[i].events;
            if (s->closed) continue;
            if (e & (EPOLLERR | EPOLLHUP)) {
                close(s->fd); reactor_unregister(&reactor, s->fd, s); free(s); continue;
            }
            if (s->listen) {
                if (e & EPOLLIN) on_accept(&reactor, s);
            } else {
                if (e & EPOLLIN) on_read(&reactor, s);
            }
        }
    }

    close(reactor.epfd);
    close(lfd);
    return 0;
}
